knimin/lib/data_access.py

```diff
@@ -2630,283 +2628,120 @@ def delete_study(self, study_id):
-                                             notes)
-                      VALUES (%s, %s, %s, %s)"""
-            sql2 = """INSERT INTO pm.study_sample (study_id, sample_id)
-                      SELECT study_ids, %s
-                      FROM unnest(%s) study_ids"""
-            for sample in samples:
-                TRN.add(sql1, [sample['id'], sample.get('is_blank', False),
-                               sample.get('barcode'), sample.get('notes')])
-                TRN.add(sql2, [sample['id'], sample['study_ids']])
-            TRN.execute()
-
-    def edit_samples(self, samples):
-        """Edits properties of samples and/or their associations with studies
-
-        Parameters
-        ----------
-        samples : list of dict
-            A list of samples specified by properties and study IDs
-            {id : str,
-                ID of the sample to edit
-             is_blank : bool, optional, default: False,
-                Whether the sample is blank
-             barcode : str, optional,
-                Assigns a barcode to the sample
-             notes : str, optional
-                Makes notes of the sample
-             study_ids : list of int, optional
-                Associates the sample with one or more studies}
-
-        Raises
-        ------
+            If the study ID does not exist
         ValueError
-            If one or more sample IDs do not exist
-            If one or more study IDs do not exist
+            If a sample is being removed and it was already plated
         """
         with TRN:
-            sample_ids = [x['id'] for x in samples]
-            self._samples_exist(sample_ids)
-            self._validate_samples(samples)
-            sql1 = """UPDATE pm.sample
-                      SET is_blank = %s, barcode = %s, notes = %s
-                      WHERE sample_id = %s"""
-            sql2 = """DELETE FROM pm.study_sample
-                      WHERE sample_id = %s"""
-            sql3 = """INSERT INTO pm.study_sample (study_id, sample_id)
-                      SELECT study_ids, %s
-                      FROM unnest(%s) study_ids"""
-            for sample in samples:
-                if any(x in sample for x in ['is_blank', 'barcode', 'notes']):
-                    TRN.add(sql1, [sample.get('is_blank', False),
-                                   sample.get('barcode'), sample.get('notes'),
-                                   sample['id']])
-                if 'study_ids' in sample:
-                    TRN.add(sql2, [sample['id']])
-                    TRN.add(sql3, [sample['id'], sample['study_ids']])
-            TRN.execute()
+            self._study_exists(study_id)
+            # Check if there are any samples that were already plated that
+            # are being removed
+            plated = set(chain.from_iterable(
+                self.get_study_plated_samples(study_id).values()))
+            samples = set(samples)
+            removed = plated - samples
```

Comment:
In general I would agree with you. However, in this case the user is not going to actively add/remove samples. This function is being used to sync the samples in LabAdmin and in Qiita, so the only thing that we receive is a set of samples, w/o distinguishing which have been added and which have been removed. Thus I find that this is the best way of doing it.